File: Study_Even_I/DataStructure/Practice_DynamicArray.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DataStructure
{
    internal class Practice_DynamicArray
    {
        class DynamicArray<T>
        {
            const int DEFAULT_SIZE = 1;
            T[] _data = new T[DEFAULT_SIZE];

            public int count;
            public int capacity
            {
                get 
                { 
                    return _data.Length; 
                }
            }

            public void Add(T item)
            {
                if(count >= capacity)
                {
                    T[]newArray = new T[count * 2];
                    for (int i = 0; i < count; i++)
                    {
                        newArray[i] = _data[i];
                    }
                    _data = newArray;
                }
                _data[count] = item;
                count++;
            }

            public T this[int index]
            {
                get
                {
                    return _data[index];
                }

                set
                {
                    _data[index] = value;
                }
            }

            public void RemoveAt(int index)
            {
                for (int i = index; i < count - 1; i++)
                {
                    _data[i] = _data[i + 1];
                }
                _data[count - 1] = default(T);
                count--;
            }

            public bool Remove(T item)
            {
                bool isFounded = false;
                for (int i = count - 1; i > - 1; i++)
                {
                    if(Comparer<T>.Default.Compare(_data[i], item) == 0)
                    {
                        RemoveAt(i);
                        isFounded = true;
                        break;
                    }   
                }
                
                return isFounded;
            }

            public void Clear()
            {
                for (int i = 0; i < count - 1; i++)
                {
                    _data[i] = default(T);
                }
                _data = new T[DEFAULT_SIZE];
                count = 0;
            }
        }

        public void DoExample()
        {
            DynamicArray<int> dynamicArray = new DynamicArray<int>();

            dynamicArray.Add(1);
            dynamicArray.Add(2);
            dynamicArray.Add(3);
            Console.WriteLine($"count : {dynamicArray.count}");
            Console.WriteLine($"last item : {dynamicArray[dynamicArray.count - 1]}");
            dynamicArray.RemoveAt(dynamicArray.count - 1);
            Console.WriteLine($"last item removed");
            Console.WriteLine($"count : {dynamicArray.count}");
            dynamicArray.Clear();
            Console.WriteLine($"Cleared");
            Console.WriteLine($"count : {dynamicArray.count}");
        }
    }
}

File: Study_Even_I/DataStructure_C/Practice_DynamicArray.c
#include <stdio.h>
#include <stdlib.h>

void Add(int item);
int Remove(int item), RemoveAt(int index);
void Clear();

static const int DEFAULT_SIZE = 1;
int* data;
int count;
static int capacity;
static int* tmpData;

int main() {
	for (int i = 0; i < 1200; i++)
	{
		Add(0);
	}

	Add(3);
	Add(5);
	Add(5);
	Add(5);
	printf("%d\n", count);
	printf("%d\n", capacity);
	for (int i = 0; i < count; i++)
	{
		printf("%d", data[i]);
	}

	Clear();
	return 0;
}

void Add(int item)
{
	if (count == 0) {
		data = (int*)malloc(DEFAULT_SIZE * sizeof(int));
		capacity = DEFAULT_SIZE;
	}
	else if (count >= capacity)
	{
		tmpData = (int*)malloc(count * 2 * sizeof(int));
		capacity = count * 2;
		for (int i = 0; i < count; i++)
			tmpData[i] = data[i];
		free(data);
		data = (int*)malloc(count * 2 * sizeof(int));
		for (int i = 0; i < count; i++)
			data[i] = tmpData[i];
		free(tmpData);
	}
	data[count] = item;
	count++;
}

int Remove(int item)
{
	int isRemoved = 0;
	for (int i = 0; i < count; i++)
	{
		if (data[i] == item) {
			isRemoved = RemoveAt(i);
			break;
		}
	}
	return isRemoved;
}

int RemoveAt(int index)
{
	int isRemoved = 0;
	if (index < count - 1) {
		for (int i = index; i < count - 1; i++)
		{
			data[i] = data[i + 1];
		}
		//data[count - 1] = NULL;
		count--;
		isRemoved = 1;
	}
	return isRemoved;
}

void Clear()
{
	free(data);
	count = 0;
	capacity = 0;
}
